### PR TITLE
HPCC-20111 LDAP Managed Scope cache improvements

### PR DESCRIPTION
**Pull request by RussWhitehead**
The refresh of the cache that holds the names of all scopes managed by LDAP
can thrash when multiple threads hit the same refresh logic. Need to
serialize the refresh of the Managed File Scopes cache. Also some minor
improvements to the debug trace logic

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

**Review by wangkx — COMMENTED** [review]
@RussWhitehead One minor change please.

**Comment by RussWhitehead** [comment]
@wangkx Please rereview

**Comment by wangkx** [comment]
@RussWhitehead looks good.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge. If we do another 6.x service release I will want to get this in it as well

**Comment by richardkchapman** [comment]
@RussWhitehead If you want it in 6.4 then please rebase and retarget this change to 6.4.24

**Comment by richardkchapman** [comment]
Also please squash to a single commit.

@ghalliday Can you sanity-check this please

File: system/security/shared/caching.cpp [hunk 1]
```diff
@@ -532,25 +532,33 @@ inline void CPermissionsCache::removeAllManagedFileScopes()
...
-        IArrayOf<ISecResource> scopes;
-        aindex_t count = m_secMgr->getManagedFileScopes(scopes);
-        if (count)
-            addManagedFileScopes(scopes);
-        m_defaultPermission = SecAccess_Unknown;//trigger refresh
-        m_lastManagedFileScopesRefresh = now;
+        CriticalBlock block(msCacheSyncCS);
+        time_t now;
+        time(&now);
+        if (0 == m_lastManagedFileScopesRefresh || m_lastManagedFileScopesRefresh < (now - m_cacheTimeout))
```
Comment (ghalliday) [inline system/security/shared/caching.cpp]: picky: would be slightly easier to read as:

    (now - m_lastManagedFileScopesRefresh) > m_cacheTimeout

**Comment by ghalliday** [comment]
@RussWhitehead looks fine to take (no need to change code with my picky comment).  Please squash then reassign to richard

**Comment by RussWhitehead** [comment]
Closing this and reopening a new one

**Pull request closed without merging**